domain/resourcegroup/runaway.go

```diff
@@ -14,5 +14,109 @@
 
 package resourcegroup
 
-// DefaultResourceGroupName is the name of the default resource group.
+import (
+	"sync/atomic"
+	"time"
+
+	rmpb "github.com/pingcap/kvproto/pkg/resource_manager"
+	"github.com/pingcap/tidb/util/dbterror/exeerrors"
+	"github.com/tikv/client-go/v2/tikv"
+	"github.com/tikv/client-go/v2/tikvrpc"
+	rmclient "github.com/tikv/pd/client/resource_group/controller"
+)
+
+// DefaultResourceGroupName is the default resource group name.
 const DefaultResourceGroupName = "default"
+
+// RunawayManager is used to detect and record runaway queries.
+type RunawayManager struct {
+	resourceGroupCtl *rmclient.ResourceGroupsController
+	// TODO: add watch records
+}
+
+// NewRunawayManager creates a new RunawayManager.
+func NewRunawayManager(resourceGroupCtl *rmclient.ResourceGroupsController) *RunawayManager {
+	return &RunawayManager{
+		resourceGroupCtl: resourceGroupCtl,
+	}
+}
+
+// DeriveChecker derives a RunawayChecker from the given resource group
+func (rm *RunawayManager) DeriveChecker(resourceGroupName string, originalSQL string, planDigest string) *RunawayChecker {
+	group, err := rm.resourceGroupCtl.GetResourceGroup(resourceGroupName)
```

Comment:
resource group are cached in resouce manager client, and return the rmpb.ResourceGroup which is more convenient

Reply:
But then you need to watch the resource group config changes in this client while info-schema is already the standard way to sync config info between all tidb servers.

Reply:
I think it's only request once at first time, then cached.

Reply:
We have considered info schema at first. As we already stored the resource group in both PD and TiDB meta, we just follow the original way to sync resource group by PD, otherwise, TiKV and Tiflash are hard to watch resource group changes.

Reply:
Though I do think storing resource groups in both PD and TiDB info schema is a little messy